File: 4.search and sorting/123.pivot in sorted/code.cpp
Ques:https://leetcode.com/problems/find-pivot-index/description/


Code:

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int pivotIndex(vector<int>& nums) {
        int total = accumulate(nums.begin(), nums.end(), 0);
        int leftSum = 0;

        for (int i = 0; i < nums.size(); ++i) {
            if (leftSum == total - leftSum - nums[i]) {
                return i;
            }
            leftSum += nums[i];
        }

        return -1;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 7, 3, 6, 5, 6};
    cout << "Pivot Index: " << sol.pivotIndex(nums) << endl;
    return 0;
}



Example Input & Output
Input:
cpp
Copy
Edit
nums = [1, 7, 3, 6, 5, 6]
Output:
Copy
Edit
3


Time & Space Complexity
Metric	Value
Time Complexity	O(n)
Space Complexity	O(1)

File: 11.heap/4. Return K largest Elements in array/code.py
Ques:https://leetcode.com/problems/kth-largest-element-in-an-array/description/

Code:

import heapq

class Solution:
    def findKthLargest(self, nums, k):
        min_heap = []

        for num in nums:
            heapq.heappush(min_heap, num)
            if len(min_heap) > k:
                heapq.heappop(min_heap)

        return min_heap[0]


if __name__ == "__main__":
    sol = Solution()
    nums = [3, 2, 3, 1, 2, 4, 5, 5, 6]
    k = 4
    print(f"The {k}th largest element is: {sol.findKthLargest(nums, k)}")


Time Complexity:
Insertion in Min-Heap: Each element is inserted into the heap, which takes O(log k) time.
Iteration through n elements: We iterate over all n elements of the array.
Pop Operation: Whenever the heap exceeds k elements, we pop the smallest element, which also takes O(log k) time.
Final Complexity Calculation:
Each insertion and pop operation costs O(log k).
We perform these operations n times.
Thus, the total complexity is O(n log k).
Space Complexity:
Min-Heap Storage: At most, k elements are stored in the heap at any time.
Ignoring Input Storage: The input array nums is given as input and does not contribute to additional space complexity.
Final Space Complexity: O(k).
Final Complexity Summary
Operation	Complexity
Time Complexity	O(n log k)
Space Complexity	O(k)